### PR TITLE
Additional classic models

### PR DESCRIPTION
**Pull request by Spiderman1013**
## PR Description

Discuss why is this PR necessary?

Closes: #64 

## What changes were made?

- High level description
Added the implementation of Gaussian Processing, Gradient Boosting, Ridge Regression, ElasticNet, AdaBoost,  and Extra Trees for classical machine learning models. This was done with a new base class that abstracted most of the manual getter/setter work done before along with repetition with passing parameters into the sklearn models. Note that Gaussian Processing and AdaBoost have an option to automatically wrap in MultiOutput with a new user set parameter in model configuration. 

In addition, two ensemble methods were added: MultiOutput and Stacking. These were done as previously designed with the classical models along with error checking for invalid models.

Reviewers: @myerspat

### CODE REVIEW

File: pyMAISE/methods/_base.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+class BaseModel(ABC):
+    def __init__(self, parameters: dict = None, default_params: dict = None):
+        """
+        Initialize the model with default and user-provided parameters.
+
+        :param parameters: Dictionary of user-provided parameters.
+        :param default_params: Dictionary of default parameters in the format: {
+            'variable_name': [default_value, type, validator(optional)] }
+        """
```
Comment (myerspat) [inline pyMAISE/methods/_base.py]: Lets get these docstrings to match the format of the other files, refer to `pyMAISE/tuner.py` or `pyMAISE/postprocessor.py` for examples.

File: pyMAISE/methods/_gaussian_process.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
+from sklearn.gaussian_process import GaussianProcessRegressor, GaussianProcessClassifier
+import pyMAISE.settings as settings
+from ._base import BaseModel
+
+
+class GaussianProcess(BaseModel):
```
Comment (myerspat) [inline pyMAISE/methods/_gaussian_process.py:6]: So the Gaussian Process should only be used when the outputs are standard scaled, somewhere whether here or in the `Tuner`/`PostProcessor`. We need to add a check that the mean is 0 and variance is 1 for the outputs.

File: docs/source/benchmarks/mit_reactor.ipynb
Comment (myerspat) [inline docs/source/benchmarks/mit_reactor.ipynb:1]: Let's finalize this benchmark so it looks presentable by finalizing the following:
- the full metrics table is printed,
- both validation and diagonal validation plots are shown in full and are not overlapping,
- include these new models in the model initialization and postprocessing discussions.

Comment (myerspat) [reply docs/source/benchmarks/mit_reactor.ipynb:1]: Also for now lets remove GP as the outputs must be standard scaled and we're min-max scaling.

File: docs/source/benchmarks/HTGR_microreactor.ipynb
Comment (myerspat) [inline docs/source/benchmarks/HTGR_microreactor.ipynb:1]: Refer to the MIT reactor comments. I'd like to get this presentable as well. We'll drop GP for now as outputs here are normalized not standard scaled.

**Review by myerspat — CHANGES_REQUESTED** [review]
@Spiderman1013, this looks really good! Thanks for picking this up. I left a few comments. After these are addressed I'll add some tasks to get the docs updated but we can wait to get that inline until after the paper is resubmitted.

**Review by myerspat — CHANGES_REQUESTED** [review]
Looking good here. There are a few things we need to do on the documentation front. All files I'm going to talk about here are in `docs/source/` and in docstrings. We need to update the supported models in `index.rst`. Also if you could go into detail in the user guide and tuner documentation found at `user_guide.rst` and `pyMAISE/tuner.py`. Mainly in the hyperparameter tuning sections. Can you also add the new models to the dictionary templates given in `models.rst`. I believe that will cover the documentation.

File: docs/source/user_guide.rst [hunk 3]
```diff
@@ -155,6 +155,11 @@ To initialize :class:`pyMAISE.Tuner`, we define each model using a list of their
 
 This indicates that we change the ``"n_estimators"`` hyperparameter to 200, the rest are initialized as default.
 
+Some models (right now just Guassian Processing), which will only work when the data is standard scaled. If still selected when not standard scaled, an error will be thrown when postprocessing the data.
+
+Also, some models also have the ability to extend to multi-output without having to pass through the multi-output object in the models array. This is done by turning `multi-output = True` in model settings.
+Currently this is supported for gradient boosting, adaboost, and stacking. Please note that when this is done, pyMAISE is actually passing the model through a ``MultiOutput`` object. The same quirks about ``MultiOutput`` Hyperparameter Tuning (as discussed in the next section) still apply.
```
Comment (myerspat) [inline docs/source/user_guide.rst:161]: Isn't support vector machine also a MultiOutput model?

File: docs/source/user_guide.rst [hunk 4]
```diff
@@ -237,6 +242,32 @@ This dictionary is then passed to the grid search tuning function:
...
+      "estimators_RF_max_depth": [5, 10, 15],
+   }
+
+For a multi-output estimator (this example has a wrapped elsatic net estimator):
+
+.. code-block:: python
+
+   multi_output_search_space = {
+      "estimator__max_depth": np.linspace(0.0001, 5, 20),
+   }
```
Comment (myerspat) [inline docs/source/user_guide.rst:267]: Why is there an additional underscore here? This is not what is shown in the paragraph above.

**Review by myerspat — CHANGES_REQUESTED** [review]
Last two comments here @CheetahJokes.

File: docs/source/user_guide.rst (hunk 4 above)
Comment (CheetahJokes) [reply docs/source/user_guide.rst:267]: This is my fault. After looking back they should all be "__" (double underscore). I will change the documentation right now.

File: docs/source/user_guide.rst (hunk 3 above)
Comment (CheetahJokes) [reply docs/source/user_guide.rst:161]: Yes, SVM doesn't natively support MultiOutput so it would make sense to include the multi-output setting to wrap our outputs. However, the SVM code is currently written the old way (without the new base class) which is why I didn't do that. Do you want me to rewrite the SVM code?

File: docs/source/user_guide.rst (hunk 4 above)
Comment (myerspat) [reply docs/source/user_guide.rst:267]: Hey @CheetahJokes, have you pushed these updates yet? I want to get this finalized.

Comment (CheetahJokes) [reply docs/source/user_guide.rst:267]: Sorry, just pushed these changes. Let me know if anything else needs to tweaked!

**Review by myerspat — APPROVED** [review]
@CheetahJokes, thanks for the hard work here! It looks good to me.
